### DISCUSSION TITLE
Can't install the wheel file with pip while offline.

### DISCUSSION THREAD

**Discussion by besdollma** [squidfunk/mkdocs-material#2591] [General]
Hi, 
I am trying to install mkdocs-material in a computer that has no access to internet.
I have download the wheel file with pip and all the dependencies.
As it seems, `mkdocs-material-extensions` is a dependency for `mkdocs-material`.
Hence I perform first `pip install mkdocs_material_extensions-1.0.1-py3-none-any.whl`, this fails since it claims `mkdocs-material` must be installed. I know people have opened prior issues on circular dependencies, and they were closed. So can you please help on how to proceed ?
Thanks.

**Comment by squidfunk** [comment]
I'm definitely not a specialist on `pip`, so maybe somebody with deeper knowledge can help here.

**Reply by facelessuser** [reply]
This is not a normal use case for me, so I don't have an answer offhand, but I am pretty sure if you are not specifically directing pip to some kind of folder to where it can find the dependencies, try to load the local, specified wheel and look online for the dependencies.

There should be a way to redirect pip from looking to PyPI to some local folder, but you have to have all the needed dependencies in that folder.  Then you should be able to install `mkdocs-material` directly and let pip auto-install dependencies from that folder.

This link looks old, but there may be some helpful info here: https://stackoverflow.com/a/11092043/3609487

**Reply by besdollma** [reply]
I will definitely try the solutions in this post in the next few days and update you. However, I have already installed this way tens of python libraries, and this is the first time I am coming across something like this (circular dependencies). Why does `mkdocs-material-extensions` require `mkdocs-material` is still a mystery to me.

**Reply by facelessuser** [reply]
It may only be compatible with a specific mkdocs-material version. I've mentioned previously that I am open to removing the dependency from that package, but I also believe the approach you are using, while maybe working in the past, may not be fully correct. But again, I don't know as I've never gone down this road before.

**Reply by besdollma** [reply]
What would you suggest as the correct approach to installing `mkdocs-material` offline?

One question:
If the only reason for the dependency is that `mkdocs-material-extensions` can work with a specific `mkdocs-material` version, isn't it enough to lock the correct version through the `mkdocs-material` dependencies?

**Reply by facelessuser** [reply]
> What would you suggest as the correct approach to installing mkdocs-material offline?

Currently, you are manually ensuring things are installed in the correct order, something `pip` resolves on its own. `pip` doesn't have an issue with this situation as it has the logic to handle this, but in your case, `pip` is offline, and when it tries to resolve things, it can't find the dependency on PyPI.  That is why you should redirect `pip` to a folder with all your wheels, and just install `mkdocs-material` letting `pip` do all the resolution. Theoretically, this should work, but I have not tried it.

The link I provided seems to suggest a way to do this, assuming the switches and such are not outdated. But I imagine the functionality should still exist even if the flags are using different names now.

> One question:
> If the only reason for the dependency is that `mkdocs-material-extensions` can work with a specific `mkdocs-material` version, isn't it enough to lock the correct version through the `mkdocs-material` dependencies?

This is an optional dependency, so it isn't installed automatically. Yes there are ways to provide optional dependencies so that you could do something like:

```py
pip install 'mkdocs-material[extras]'
```

But `mkdocs-material` would have to decide to go down this road.

**Reply by facelessuser** [reply]
Does material include `mkdocs-material-extensions` by default? I can't remember, my earlier statement about being an optional dependency maybe can be ignored if it does....

**Reply by squidfunk** [reply]
Yep:

https://github.com/squidfunk/mkdocs-material/blob/9296dc3bcabe33a1a3620553b665c0216f571a59/requirements.txt#L26

**Reply by facelessuser** [reply]
Okay, in that case, we could let Material fully control the version to use. Currently, it is not incorrect to state there is a circular dependency though as `mkdocs-material-extentsions` does reference material. `pip` does resolve this without issue (if used appropriately). With that said, I am open to removing the specified dependency from `mkdocs-material-extentsions`, but I don't think it is expressly needed if `pip` is used appropriately, but obviously, I have not verified the suggested solution regarding `pip`.

**Reply by besdollma** [reply]
I will definitely try and explore more `pip` flags to use it as you suggested. I will update as soon as I get the chance to do that. 

However, if I am allowed to give my opinion in the matter, even if `pip` resolves this, the circular dependency issue remains. Tools like `pipdeptree`, `pdoc`, `pyinstaller` and other tools that do package analysis/ops with all the dependencies might occur difficulties and work partially if not at all. 

The issue is also theoretical in a manner of speaking, if `mkdocs-material` requires `mkdocs-material-extensions`, this is not really an extension since it is required. If `mkdocs-material-extensions` is required and is a dependency of `mkdocs-material` then it really isn't an extension, is it?

**Reply by facelessuser** [reply]
> The issue is also theoretical in a manner of speaking, if `mkdocs-material` requires `mkdocs-material-extensions`, this is not really an extension since it is required. If `mkdocs-material-extensions` is required and is a dependency of `mkdocs-material` then it really isn't an extension, is it?

Sure it is. They are 3rd party extensions. And it actually makes perfect sense that they would be locked to a specific version of Material. What makes it more complicated is that Material then installs them as a requirement.

For example, this is no different than `pymdown-extensions` which are extensions for Python Markdown. They are locked to a minimum version of Python Markdown because they don't work with versions before that.

As an extension maintainer, if Material updates and changes where and how they provide their icons, I have the ability to specify which versions I support, and the user can pin `mkdocs-material` until my extensions are updated to support the latest. Then I can release the next version which depends on the latest without breaking users on older versions.

So, IMHO, it actually makes way more sense for the extensions to be pinned to a specific Material version than it does for Material to be pinned to an extension version. With that said, I'm pretty sure @squidfunk would be pretty careful in making sure to coordinate with me to ensure there is little to no breakage. So maybe it makes sense to assume that no one would manually specify the extensions, and they can be assumed to be a piece of Material that is broken out into a separate package maintained by a separate individual (namely me).

> However, if I am allowed to give my opinion in the matter, even if `pip` resolves this, the circular dependency issue remains. Tools like `pipdeptree`, `pdoc`, `pyinstaller` and other tools that do package analysis/ops with all the dependencies might occur difficulties and work partially if not at all.

It _might_ be problematic with some of the above. I have not confirmed at all. I know the first reference (`pipdeptree`) can output warnings for circular dependencies which makes it sound like it'll still work.

---

Again, I'm open to removing the circular dependency, but I'd argue there is nothing "technically" wrong with what is being done. Personally, even if we do remove the circular dependency, adjusting how you are loading wheels will probably be easier for you to maintain in the long term as then you can just install like you would with pip instead of managing dependency order loading.

**Reply by facelessuser** [reply]
In summary. I have always viewed the package I maintain as an extension, but due to the fact that Material includes them as a dependency, that mindset may actually be incorrect. It may make more sense to view them as a separate piece of `mkdocs-material` that is maintained by me.

**Reply by facelessuser** [reply]
What are your opinions on this @squidfunk? Do you think we should drop the dependency on the extension pack and just let Material define what it wants?

**Reply by squidfunk** [reply]
Maybe it's worth reconsidering moving the extensions back into this repository? IIRC we moved it out of this repository because it was not possible to get a development environment working without installing the project itself with `pip install .`, but looking back that shouldn't be too big of an investment. I've some ideas for Material-specific MkDocs plugins anyway, which I guess would require the same step. Any thoughts on that idea?

**Reply by facelessuser** [reply]
I'm open to it.  I think either way works, but it is more what makes sense to you.

If we kept it separate, you could continue to install it separately and avoid the `pip install .`. `mkdocs-material` would have to strictly control which version(s) it supports.

If you pull it in, you may need to `pip install .` or use [`venv`](https://docs.python.org/3/tutorial/venv.html) environments to test. Neither of which bothers me.

If you are heading down the road of adding additional python plugins anyway, then I don't see an issue with you pulling in this one.

**Reply by besdollma** [reply]
I can confirm that I was able to install based on the suggestions here https://stackoverflow.com/a/11092043/3609487.

**Reply by facelessuser** [reply]
Cool, that is what I figured. In the long run, that will make this issue moot, but with that said, there is at least a possibility that we may change how dependencies reference one another, but that shouldn't affect you anymore.

**Reply by rafmagns-skepa-dreag** [reply]
I've run into the same issue, but when running bazel instead of installing offline. I wish there were another way to resolve this but unfortunately bugs filed against bazel rules have been marked as "won't fix. desired behavior". If there's anything you'd like me to do to help out, please let me know.

**Reply by facelessuser** [reply]
@rafmagns-skepa-dreag My stance is pretty much the same as it has always been. I'm open to doing away with the circular dependency and/or moving the extensions directly into Material.  I haven't yet run into a situation where it was absolutely necessary, but maybe you have a compelling reason for us to reconsider?

**Reply by rafmagns-skepa-dreag** [reply]
My only compelling reason is that, while pip will allow it, other build tools explicitly disallow circular dependencies. Google's tool, [Bazel](https://bazel.build/), is what I ran up against. As I said above, other [bugs filed about circular dependencies to the Bazel rules](https://github.com/bazelbuild/rules_python/issues/411) have been marked as "won't fix; desired behavior". Fixing this issue is unfortunately absolutely necessary for me to continue to use these two pacakges. I know it's more work/overhead for you two maintainers to go this way so I understand if supporting pip is where you want to draw the line.

**Reply by squidfunk** [reply]
The canonical way to use and install this project is `pip`, which resolves circular dependencies successfully. We can move the extensions code back into this repository to remove the circular dependency, but this is currently not a priority.

**Reply by facelessuser** [reply]
I approached this originally as an extension, not an integral part of Material. It was not desired to be placed in Material, so I treated it as an optional extra. I had thought that people would install it if they wanted the added benefit. That is why the extensions had a dependency added specifying a given Material version. What I thought was as an external extension has become more of an official module and an important piece of the main project. It doesn't ship without the extension.

It has become apparent that Material won't move forward breaking this extension's behavior without communication as it depends on this functionality. So I think letting Material manage versions of this plugin and such makes more sense than me managing it in the reverse.

Bazel's choice to support or not support circular dependencies is probably a much lower reason on my list as I am more concerned with what is officially supported.

Personally, I kind of think it makes sense to let Material absorb this extension since it won't ship without it anyways. That had always been my preferred choice, but at the time, @squidfunk was worried about supporting Python code in addition to the browser end. That may have changed now.

So, what I'll do is go ahead and remove the dependency. I will probably add an external check to throw an error if it is run with an unsupported Material version. If Material chooses to absorb the plugin, I'll be happy to help it do so.

**Reply by rafmagns-skepa-dreag** [reply]
Thank you for the resolution! You two have been very responsive and I really appreciate it. Thanks for your time and energy both on this issue and the packages.

**Reply by facelessuser** [reply]
I'll try to get the package updated no later than this weekeend.

**Reply by facelessuser** [reply]
https://github.com/facelessuser/mkdocs-material-extensions/releases/tag/1.0.2 has been tagged for release. Should be on PyPI shortly. I think this resolves everyone's issues.
